// @pulumi/pulumi/config.js
"use strict";
// Copyright 2016-2018, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
const util = require("util");
const errors_1 = require("./errors");
const log = require("./log");
const runtime = require("./runtime");
/**
 * Config is a bag of related configuration state.  Each bag contains any number of configuration variables, indexed by
 * simple keys, and each has a name that uniquely identifies it; two bags with different names do not share values for
 * variables that otherwise share the same key.  For example, a bag whose name is `pulumi:foo`, with keys `a`, `b`,
 * and `c`, is entirely separate from a bag whose name is `pulumi:bar` with the same simple key names.  Each key has a
 * fully qualified names, such as `pulumi:foo:a`, ..., and `pulumi:bar:a`, respectively.
 */
class Config {
    constructor(name) {
        // To ease migration of code that already does new Config("<package>:config"), treat this as if
        // just new Config("<package>") was called.
        if (name.endsWith(":config")) {
            const newName = name.replace(/:config$/, "");
            log.warn(util.format("`:config` is no longer required at the end of configuration " +
                "bag names and support will be removed in a future version, please " +
                "use new Config(\"%s\") instead.", newName));
            name = newName;
        }
        this.name = name;
    }
    /**
     * get loads an optional configuration value by its key, or undefined if it doesn't exist.
     *
     * @param key The key to lookup.
     */
    get(key) {
        return runtime.getConfig(this.fullKey(key));
    }
    /**
     * getBoolean loads an optional configuration value, as a boolean, by its key, or undefined if it doesn't exist.
     * If the configuration value isn't a legal boolean, this function will throw an error.
     *
     * @param key The key to lookup.
     */
    getBoolean(key) {
        const v = this.get(key);
        if (v === undefined) {
            return undefined;
        }
        else if (v === "true") {
            return true;
        }
        else if (v === "false") {
            return false;
        }
        throw new ConfigTypeError(this.fullKey(key), v, "boolean");
    }
    /**
     * getNumber loads an optional configuration value, as a number, by its key, or undefined if it doesn't exist.
     * If the configuration value isn't a legal number, this function will throw an error.
     *
     * @param key The key to lookup.
     */
    getNumber(key) {
        const v = this.get(key);
        if (v === undefined) {
            return undefined;
        }
        const f = parseFloat(v);
        if (isNaN(f)) {
            throw new ConfigTypeError(this.fullKey(key), v, "number");
        }
        return f;
    }
    /**
     * getObject loads an optional configuration value, as an object, by its key, or undefined if it doesn't exist.
     * This routine simply JSON parses and doesn't validate the shape of the contents.
     *
     * @param key The key to lookup.
     */
    getObject(key) {
        const v = this.get(key);
        if (v === undefined) {
            return undefined;
        }
        try {
            return JSON.parse(v);
        }
        catch (err) {
            throw new ConfigTypeError(this.fullKey(key), v, "JSON object");
        }
    }
    /**
     * require loads a configuration value by its given key.  If it doesn't exist, an error is thrown.
     *
     * @param key The key to lookup.
     */
    require(key) {
        const v = this.get(key);
        if (v === undefined) {
            throw new ConfigMissingError(this.fullKey(key));
        }
        return v;
    }
    /**
     * requireBoolean loads a configuration value, as a boolean, by its given key.  If it doesn't exist, or the
     * configuration value is not a legal boolean, an error is thrown.
     *
     * @param key The key to lookup.
     */
    requireBoolean(key) {
        const v = this.getBoolean(key);
        if (v === undefined) {
            throw new ConfigMissingError(this.fullKey(key));
        }
        return v;
    }
    /**
     * requireNumber loads a configuration value, as a number, by its given key.  If it doesn't exist, or the
     * configuration value is not a legal number, an error is thrown.
     *
     * @param key The key to lookup.
     */
    requireNumber(key) {
        const v = this.getNumber(key);
        if (v === undefined) {
            throw new ConfigMissingError(this.fullKey(key));
        }
        return v;
    }
    /**
     * requireObject loads a configuration value, as a number, by its given key.  If it doesn't exist, or the
     * configuration value is not a legal number, an error is thrown.
     *
     * @param key The key to lookup.
     */
    requireObject(key) {
        const v = this.getObject(key);
        if (v === undefined) {
            throw new ConfigMissingError(this.fullKey(key));
        }
        return v;
    }
    /**
     * fullKey turns a simple configuration key into a fully resolved one, by prepending the bag's name.
     *
     * @param key The key to lookup.
     */
    fullKey(key) {
        return `${this.name}:${key}`;
    }
}
exports.Config = Config;
/**
 * ConfigTypeError is used when a configuration value is of the wrong type.
 */
class ConfigTypeError extends errors_1.RunError {
    constructor(key, v, expectedType) {
        super(`Configuration '${key}' value '${v}' is not a valid ${expectedType}`);
    }
}
/**
 * ConfigMissingError is used when a configuration value is completely missing.
 */
class ConfigMissingError extends errors_1.RunError {
    constructor(key) {
        super(`Missing required configuration variable '${key}'\n` +
            `\tplease set a value using the command \`pulumi config set ${key} <value>\``);
        this.key = key;
    }
}
